pkg/strategy/drift/strategy.go

```diff
@@ -0,0 +1,363 @@
+
+	"github.com/c9s/bbgo/pkg/bbgo"
+	"github.com/c9s/bbgo/pkg/fixedpoint"
+	"github.com/c9s/bbgo/pkg/indicator"
+	"github.com/c9s/bbgo/pkg/types"
+	"github.com/c9s/bbgo/pkg/util"
+)
+
+const ID = "drift"
+
+var log = logrus.WithField("strategy", ID)
+
+func init() {
+	bbgo.RegisterStrategy(ID, &Strategy{})
+}
+
+type Strategy struct {
+	Symbol string `json:"symbol"`
+
+	bbgo.StrategyController
+	types.Market
+	types.IntervalWindow
+
+	*bbgo.Environment
+	*types.Position
+	*types.ProfitStats
+	*types.TradeStats
+
+	drift    *indicator.Drift
+	atr      *indicator.ATR
+	midPrice fixedpoint.Value
+	lock     sync.RWMutex
+
+	Stoploss   fixedpoint.Value `json:"stoploss"`
+	CanvasPath string           `json:"canvasPath"`
+
+	ExitMethods bbgo.ExitMethodSet `json:"exits"`
+	Session     *bbgo.ExchangeSession
+	*bbgo.GeneralOrderExecutor
+	*bbgo.ActiveOrderBook
+}
+
+func (s *Strategy) ID() string {
+	return ID
+}
+
+func (s *Strategy) InstanceID() string {
+	return fmt.Sprintf("%s:%s", ID, s.Symbol)
+}
+
+func (s *Strategy) Subscribe(session *bbgo.ExchangeSession) {
+	session.Subscribe(types.KLineChannel, s.Symbol, types.SubscribeOptions{
+		Interval: s.Interval,
+	})
+	session.Subscribe(types.KLineChannel, s.Symbol, types.SubscribeOptions{
+		Interval: types.Interval1m,
+	})
+
+	if !bbgo.IsBackTesting {
+		session.Subscribe(types.MarketTradeChannel, s.Symbol, types.SubscribeOptions{})
+	}
+	s.ExitMethods.SetAndSubscribe(session, s)
+}
+
+var Three fixedpoint.Value = fixedpoint.NewFromInt(3)
+var Two fixedpoint.Value = fixedpoint.NewFromInt(2)
+
+func (s *Strategy) GetLastPrice() (lastPrice fixedpoint.Value) {
+	var ok bool
+	if s.Environment.IsBackTesting() {
+		lastPrice, ok = s.Session.LastPrice(s.Symbol)
+		if !ok {
+			log.Error("cannot get lastprice")
+			return lastPrice
+		}
+	} else {
+		s.lock.RLock()
+		if s.midPrice.IsZero() {
+			lastPrice, ok = s.Session.LastPrice(s.Symbol)
+			if !ok {
+				log.Error("cannot get lastprice")
+				return lastPrice
+			}
+		} else {
+			lastPrice = s.midPrice
+		}
+		s.lock.RUnlock()
+	}
+	return lastPrice
+}
+
+var Delta fixedpoint.Value = fixedpoint.NewFromFloat(0.01)
+
+func (s *Strategy) ClosePosition(ctx context.Context) (*types.Order, bool) {
+	order := s.Position.NewMarketCloseOrder(fixedpoint.One)
+	if order == nil {
+		return nil, false
+	}
+	order.TimeInForce = ""
+	balances := s.Session.GetAccount().Balances()
+	baseBalance := balances[s.Market.BaseCurrency].Available
+	price := s.GetLastPrice()
+	if order.Side == types.SideTypeBuy {
+		quoteAmount := balances[s.Market.QuoteCurrency].Available.Div(price)
+		if order.Quantity.Compare(quoteAmount) > 0 {
+			order.Quantity = quoteAmount
+		}
+	} else if order.Side == types.SideTypeSell && order.Quantity.Compare(baseBalance) > 0 {
+		order.Quantity = baseBalance
+	}
+	for {
+		if s.Market.IsDustQuantity(order.Quantity, price) {
+			return nil, true
+		}
+		createdOrders, err := s.GeneralOrderExecutor.SubmitOrders(ctx, *order)
```

Comment:
the s.GeneralOrderExecutor provides ClosePosition method , I think you can use it?

Reply:
There are additional checks done in my ClosePosition method, plus the ClosePosition in GeneralOrderExecutor cannot get correct latest price.
my ClosePosition method will also retry closing the position in the loop until position became dust or zero.